### ISSUE TITLE
Improve performance with parallelization when artifact compression is none

### ISSUE THREAD

**Issue by dseapy** [argoproj/argo-workflows#4014]
# Summary

When using compression of "none" on artifacts with 10000 files the get or put takes tens of minutes.  Adjusting the resources on the executor doesn't seem to help.  A quick look at pkg, and I believe this is because the s3client is going file by file sequentially instead of having a pool of workers running through the files in parallel.

# Use Cases

Uploading many files as an artifact for a subsequent highly parallel step, and also then joining the results after the step is complete.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
@dseapy you know the code I think? Would you like to submit a PR? You could use a `WaitGroup`, or see if there's a better way to do this?

https://github.com/argoproj/pkg/blob/master/s3/s3.go#L175

**Comment by dseapy** [comment]
Yeah, I can probably work on this but given my project's current deadlines it might be a few weeks before I can get to it.

**Comment by NikeNano** [comment]
Is this still relevant @alexec, I would be happy to take a look at this!

**Comment by alexec** [comment]
This is not a popular proposal, so I don't think the core team will implement this soon. If you're looking for something popular to work on, you can find issues with many thumbs-up.

**Comment by NikeNano** [comment]
Will take a look at the more popular issues to start with, thanks!

**Comment by DocX** [comment]
We have exactly the same issue. And I found someone else reporting similar issue in https://github.com/argoproj/argo-workflows/issues/4014

Is there any roadmap for this? Instead I may be able to look at it. Can someone point me to relevant places in argo?

If I understand it correctly argo uses `minio` client for s3?

**Comment by ldejuan** [comment]
Since I was stuck with this issue, I decided to use a minio client on my docker image to copy the artfacts back to the S3, against the argo design' concepts. In my code I had to explicity pass the bucket key and not in the workflow manifest. And the performances were incredible. When running a workflow with a lot of small artifacts, argo underperforms dramatically. This issue should be addressed.

**Comment by DocX** [comment]
Yes, we also resorted to just use `mc` CLI client. It means changing the docker image we use and adding bash script that runs it. It is much faster, since the `mc` CLI uses parallelisation.

For reference, this is script we did:

```sh
mc alias set s3 https://s3.amazonaws.com $S3_AWS_KEY $S3_AWS_SECRET
mc cp --recursive /tmp/logs/ s3/$S3_BUCKET_NAME/$S3_TARGET_KEY/logs/
```

**Comment by DocX** [comment]
@alexec I've put together simple PR, would you be able to review it?

**Comment by djmcgreal-cc** [comment]
@DocX, @alexec  - how did this end up?

I'm now re-writing my in-development workflow to use rclone instead.

:(

**Comment by DocX** [comment]
we use custom container with `mc` instead of the built in Argo artifact upload
